documentation/writing_tests.md

```diff
@@ -0,0 +1,949 @@
+
+- The titles sent to the `describe` calls should represent what the collection of its inner test cases (`it` calls) are testing.
+- The titles sent to each `it` call should represent what each specific test case is testing.
+
+Therefore, `describe` titles should be written to reflect something that each of its `it` test cases (or inner `describe` blocks) share in common.
+
+In addition to being expressive, test titles might have specific hashtags (as in, words preceded by the pound sign), to facilitate pattern matching.
+
+#### Pattern matching test titles
+
+`mocha` allows us to only run certain tests as long as they match a given query, with the following arguments:
+
+```
+Test Filters
+  --fgrep, -f   Only run tests containing this string                   [string]
+  --grep, -g    Only run tests matching this string or regexp           [string]
+  --invert, -i  Inverts --grep and --fgrep matches                     [boolean]
+```
+
+To be able to filter tests through pattern matching with any of these filters, the `describe` and `it` titles can have any hashtag, but preferably one or more of the following:
+
+- `#browser` for tests that are only expected to run in the browser.
+- `#node` for tests that are only expected to run on NodeJS.
+- `#live` for tests that should only run live and not be recorded, nor played back.
+
+### Inner parts of a test
+
+While writing test cases, it will be easier to think of the test case's contents as if there were three clear divisions:
+
+- The first lines of the test should deal with setting up the necessary variables and values that will be used for the main functionality that will be tested.
+- The body of the test should focus on using these values to trigger the main functionality that is being tested.
+- The last set of lines should be focused on comparing the results obtained, to confirm that they are in the expected shape, or contain the expected values.
+
+### Using conditionals
+
+Ideally, each test case should only have one possible behavior. To minimize unexpected outcomes, conditionals should be avoided. The only case in which conditionals are encouraged is to specify when to skip a test. Some tests might only run in a browser, or only run in NodeJS, or only run live (no record or playback), for this purpose, conditionals can be used. For example:
+
+```ts
+import { isNode } from "@azure/core-http";
+
+describe("testing the client's basic methods", function() {
+  let client: Client;
+
+  beforeEach(function() {
+    client = new Client();
+  });
+
+  it("should test A #browser", function() {
+    if (isNode) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+
+  it("should test B #node", function() {
+    if (!isNode) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+
+  it("should test C #live", function() {
```

Comment:
Just trying to understand your system here - are you saying that any test that doesn't say #live will activate recordings and run during the unit-test phase?

Reply:
Also, does this imply that #live tests always run in both the browser and in node? Or do I combine flags?

Reply:
This is described in more detail a bit before this snippet. I mean here: https://github.com/Azure/azure-sdk-for-js/pull/7784/files/870108e6e0b34e20f65796a0bd379c3d774e1aa0#diff-a43d448b275343b7d964b6e68f7cabf4R840

Hashtags are optional though. One "can" use them.